components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -3194,11 +2903,662 @@ subroutine rain_immersion_freezing(t,    &
+   sources = qr + (qcaut+qcacc+qimlt+qcshd)*dt
+   if (sinks.gt.sources .and. sinks.ge.1.e-20_rtype) then
+      ratio  = sources/sinks
+      qrevp  = qrevp*ratio
+      qrcol  = qrcol*ratio
+      qrheti = qrheti*ratio
+   endif
+
+end subroutine rain_water_conservation 
+
+subroutine ice_water_conservation(qitot,qidep,qinuc,qiberg,qrcol,qccol,qrheti,qcheti,dt,    &
+   qisub,qimlt) 
+
+   implicit none 
+
+   real(rtype), intent(in) :: qitot, qidep, qinuc, qrcol, qccol, qrheti, qcheti, qiberg, dt
+   real(rtype), intent(inout) :: qisub, qimlt 
+   real(rtype) :: sinks, sources, ratio 
+
+   sinks   = (qisub+qimlt)*dt
+   sources = qitot + (qidep+qinuc+qrcol+qccol+  &
+        qrheti+qcheti+qiberg)*dt
+   if (sinks.gt.sources .and. sinks.ge.1.e-20_rtype) then
+      ratio = sources/sinks
+      qisub = qisub*ratio
+      qimlt = qimlt*ratio
+   endif
+
+
+end subroutine ice_water_conservation 
+
+
+subroutine update_prognostic_ice(qcheti,qccol,qcshd,    & 
+   nccol,ncheti,ncshdc,    & 
+   qrcol,nrcol,qrheti,nrheti,nrshdr,    & 
+   qimlt,nimlt,qisub,qidep,qinuc,ninuc,nislf,nisub,qiberg,    & 
+   exner,xxls,xlf,    & 
+   log_predictNc,log_wetgrowth,dt,nmltratio,rhorime_c,    &
+   th,qv,qitot,nitot,qirim,birim,qc,nc,qr,nr)
+
+   !-- ice-phase dependent processes:
+   implicit none 
+
+   real(rtype), intent(in) :: qcheti
+   real(rtype), intent(in) :: qccol
+   real(rtype), intent(in) :: qcshd
+   real(rtype), intent(in) :: nccol 
+   real(rtype), intent(in) :: ncheti
+   real(rtype), intent(in) :: ncshdc  
+
+   real(rtype), intent(in) :: qrcol 
+   real(rtype), intent(in) :: nrcol 
+   real(rtype), intent(in) :: qrheti 
+   real(rtype), intent(in) :: nrheti 
+   real(rtype), intent(in) :: nrshdr 
+
+   real(rtype), intent(in) :: qimlt 
+   real(rtype), intent(in) :: nimlt 
+   real(rtype), intent(in) :: qisub
+   real(rtype), intent(in) :: qidep 
+   real(rtype), intent(in) :: qinuc 
+   real(rtype), intent(in) :: ninuc 
+   real(rtype), intent(in) :: nislf 
+   real(rtype), intent(in) :: nisub 
+   real(rtype), intent(in) :: qiberg 
+   real(rtype), intent(in) :: exner 
+   real(rtype), intent(in) :: xlf 
+   real(rtype), intent(in) :: xxls 
+
+   logical, intent(in) :: log_predictNc
+   logical, intent(in) :: log_wetgrowth
+   real(rtype), intent(in) :: dt 
+   real(rtype), intent(in) :: nmltratio 
+   real(rtype), intent(in) :: rhorime_c 
+
+   real(rtype), intent(inout) :: th 
+   real(rtype), intent(inout) :: qv 
+   real(rtype), intent(inout) :: qc
+   real(rtype), intent(inout) :: nc 
+   real(rtype), intent(inout) :: qr 
+   real(rtype), intent(inout) :: nr 
+   real(rtype), intent(inout) :: qitot 
+   real(rtype), intent(inout) :: nitot 
+   real(rtype), intent(inout) :: qirim 
+   real(rtype), intent(inout) :: birim 
+
+   real(rtype) :: dum 
+
+   qc = qc + (-qcheti-qccol-qcshd-qiberg)*dt
+   if (log_predictNc) then
+      nc = nc + (-nccol-ncheti)*dt
+   endif
+
+   qr = qr + (-qrcol+qimlt-qrheti+            &
```

Comment:
I guess we could clean up these lines a bit while we are at it.  Like in this case, there really isn't a reason to have this line be two lines.